File: static/js/ChessPieces/Pawn.js
class Pawn extends ChessPiece{

	constructor(){
		super();
		this.opponentPlaceIsSetEnPassant = false;
		this.opponentPlaceEnPassant = null;
		this.placeIsSetEnPassant = false;
		this.placeEnPassant = null;
	}

	movablePlaces(
		compPawnStartingPositions, 
		currentEnPassantPlaceId,
		enPassantOpponentLeft, 
		enPassantOpponentRight, 
		isEnPassant, 
		currentEnPassantOpponentPlaceId, 
		x, 
		y
	){
		
		// console.log("pawn: "+this.getPawnMovablePlaces(x, y)+" "+x+" "+y);
		// console.log("shrunken pawn: "+this.shrinkPawnArray(this.getPawnMovablePlaces(x, y), "playing"));
		return this.enPassantMovement(
			compPawnStartingPositions, 
			currentEnPassantPlaceId,
			enPassantOpponentLeft, 
			enPassantOpponentRight, 
			isEnPassant, 
			currentEnPassantOpponentPlaceId,
			this.shrinkPawnArray(this.getPawnMovablePlaces(x, y), "playing"),
			x,
			y
		);
	}

    getEnPassantPlace(){
    	if(this.placeIsSetEnPassant){
    		this.placeIsSetEnPassant = true;
    		return this.placeEnPassant;
    	}
    	return undefined;
    }

     getEnPassantOpponentPlace(){
    	if(this.opponentPlaceIsSetEnPassant){
    		this.opponentPlaceIsSetEnPassant = false;
    		return this.opponentPlaceEnPassant;
    	}
    	return undefined;
    }

	getPawnMovablePlaces(x, y){
		var matrix = this.live_chessboard_matrix;
		var placeIds = [];
		if(x>=0&&x<=7&&y>=0&&y<=7){
			var leftFwd1Element = document.getElementById(this.id_gen(y-1, x-1));
			var rightFwd1Element = document.getElementById(this.id_gen(y-1, x+1));
			var fwd1Element = document.getElementById(this.id_gen(y-1, x));
			var fwd2Element = document.getElementById(this.id_gen(y-2, x));
			// console.log("Id gen: "+
			// 			this.id_gen(y-1, x-1)+
			// 			" "+
			// 			this.id_gen(y-1, x+1)+
			// 			" "+
			// 			this.id_gen(y-1, x)+
			// 			" "+
			// 			this.id_gen(y-2, x)
			// );
			if(leftFwd1Element!=null){
				if(leftFwd1Element.childElementCount!=0){
					var leftFwd1Id = leftFwd1Element.firstElementChild.id;
					if(this.isType(leftFwd1Id, "comp_")){
						placeIds[0] = leftFwd1Element.id;
					}
				}
			}
			if(rightFwd1Element!=null){
				if(rightFwd1Element.childElementCount!=0){
					var rightFwd1Id = rightFwd1Element.firstElementChild.id;
					if(this.isType(rightFwd1Id, "comp_")){
						placeIds[1] = rightFwd1Element.id;
					}
				}
			}
			if(fwd1Element!=null){
				if(fwd1Element.childElementCount==0){
					placeIds[2] = fwd1Element.id;
				}
			}
			if(fwd2Element!=null){
				if(fwd2Element.childElementCount==0){
					placeIds[3] = fwd2Element.id;
				}
			}
		}
		return placeIds;
	}

	attackingPlaces(x, y){
		var attackingPawnPlaces = this.shrinkPawnArray(this.getPawnMovablePlaces(x, y), "checking");
		var new_array = [];
		var first = document.getElementById(attackingPawnPlaces[0]);
		if(first!=null){
			if(first.firstElementChild!=null){
				var next = first.firstElementChild; 
				if(this.isType(next.id, "comp_pawn")){
					new_array.push(attackingPawnPlaces[0]);
				}		
			}
		}
		var second = document.getElementById(attackingPawnPlaces[1]);
		if(second!=null){
			if(second.firstElementChild!=null){
				var next = first.firstElementChild; 
				if(this.isType(next.id, "comp_pawn")){
					new_array.push(attackingPawnPlaces[0]);
				}		
			}
		}
		return new_array;
	}

	shrinkPawnArray(array, mechanic_needed){
		var new_array = [];
		if(mechanic_needed=="playing"){
			if(array[0]!=null){
				if(array[0]!=""){
					new_array.push(array[0]);
				}
			}
			if(array[1]!=null){
				if(array[1]!=""){
					new_array.push(array[1]);
				}
			}
			if(array[2]!=null){
				if(array[2]!=""){
					new_array.push(array[2]);
				}
			}
			if(array[3]!=null){
				if(array[3]!=""){
					new_array.push(array[3]);
				}
			}
		} else {
			if(array[0]!=""){
				var leftElement = document.getElementById(array[0]);
				if(leftElement!=null){
					if(leftElement.firstElementChild!=null){
						piece = leftElement.firstElementChild;
						if(this.isType(piece.id, "comp_pawn")){
							new_array.push(array[0]);
						}
					}
				}
			}
			if(array[1]!=""){
				var rightElement = document.getElementById(array[1]);
				if(rightElement!=null){
					if(rightElement.firstElementChild!=null){
						piece = rightElement.firstElementChild;
						if(this.isType(piece.id, "comp_pawn")){
							new_array.push(array[1]);
						}
					}
				}
			}
		}
		return new_array;
	}

	enPassantMovement(
		compPawnStartingPositions, 
		currentEnPassantPlaceId, 
		enPassantOpponentLeft, 
		enPassantOpponentRight, 
		isEnPassant, 
		currentEnPassantOpponentPlaceId, 
		rookArray, 
		x,
		y 
	){
		var newArray = [];
		var pawnHasLeft = false;
		var pawnHasRight = false;
		var leftOfPawn = this.id_gen(y, x-1);
		var rightOfPawn = this.id_gen(y, x+1);
		var leftOfPawnElement = document.getElementById(leftOfPawn);
		var rightOfPawnElement = document.getElementById(rightOfPawn);
		if(leftOfPawnElement!=null){
			if(leftOfPawnElement.firstElementChild!=null){
				var enPassantSpace = this.pawnReadyEnPassant(
					compPawnStartingPositions, 
					currentEnPassantPlaceId,
					leftOfPawnElement.firstElementChild.id, 
					leftOfPawn
				);
				// console.log("Curr en pass move: "+currentEnPassantPlaceId);
				// console.log("Pawn ready: "+enPassantSpace);
				if(enPassantSpace!=""){
					rookArray.push(enPassantSpace);
					enPassantOpponentLeft = leftOfPawnElement.firstElementChild.id;
					currentEnPassantOpponentPlaceId = leftOfPawn;
					isEnPassant = true;
					this.opponentPlaceIsSetEnPassant = true;
					this.opponentPlaceEnPassant = leftOfPawn;
					// console.log("En Passant left: "+enPassantSpace);
				}
			}
		}

		if(rightOfPawnElement!=null){
			if(rightOfPawnElement.firstElementChild!=null){
				var enPassantSpace = this.pawnReadyEnPassant(
					compPawnStartingPositions, 
					currentEnPassantPlaceId,
					rightOfPawnElement.firstElementChild.id, 
					rightOfPawn
				);
				if(enPassantSpace!=""){
					rookArray.push(enPassantSpace);
					enPassantOpponentRight = rightOfPawnElement.firstElementChild.id;
					currentEnPassantOpponentPlaceId = rightOfPawn;
					isEnPassant = true;
					this.opponentPlaceIsSetEnPassant = true;
					this.opponentPlaceEnPassant = rightOfPawn;
					// console.log("En Passant right: "+enPassantSpace);
				}
			}
		}
		return rookArray;
	}

	pawnReadyEnPassant(compPawnStartingPositions, currentEnPassantPlaceId, pieceId, placeId){
		var newPlaceId = "";
		// console.log("PieceId: "+pieceId);
		if(this.isType(pieceId, "comp_pawn")){
			for(var i=0; i<compPawnStartingPositions.length; i++){
				if(this.isType(pieceId, String(i+1))){
					var posBefore = this.findPlaceCoordinates(compPawnStartingPositions[i]);
					var y = posBefore[0] + 1;
					var x = posBefore[1];
					var posNow = this.findPlaceCoordinates(placeId);
					var nY = posNow[0] - 1;
					var nX = posNow[1];
					var placeIdWithPosBefore = this.id_gen(y, x);
					var placeIdWithPosNow = this.id_gen(nY, nX);
					// console.log("Pos en passant: "+placeIdWithPosBefore+" "+placeIdWithPosNow);
					if(placeIdWithPosBefore==placeIdWithPosNow){
						newPlaceId = placeIdWithPosNow;
						currentEnPassantPlaceId = placeIdWithPosNow;
						this.placeIsSetEnPassant = true;
						this.placeEnPassant = placeIdWithPosNow;
						// console.log("curr en pass: "+currentEnPassantPlaceId);
					}
				}
			}
		}
		return newPlaceId;
	}

}

File: chess_app/ChessEngine/ChessPieces/Pawn.py
from ChessPiece import ChessPiece
from chess_app.ChessEngine.Selecter import Select


class Pawn(ChessPiece):

    def __init__(self):
        super(Pawn, self).__init__()
        self.opponentPlaceIsSetEnPassant = False
        self.opponentPlaceEnPassant = None
        self.placeIsSetEnPassant = False
        self.placeEnPassant = None

    def movablePlaces(self, compPawnStartingPositions, currentEnPassantPlaceId, enPassantOpponentLeft, enPassantOpponentRight, isEnPassant, currentEnPassantOpponentPlaceId, x, y):

        newArray = self.enPassantMovement(
            compPawnStartingPositions,
            currentEnPassantPlaceId,
            enPassantOpponentRight,
            enPassantOpponentLeft,
            isEnPassant,
            currentEnPassantOpponentPlaceId,
            self.shrinkPawnArray(self.getPawnMovablePlaces(x, y), "playing"),
            x,
            y
        )

        return newArray

    # gets the postion of the pawns en passant position if it has been set
    def getEnPassantPlace(self):
        if self.placeIsSetEnPassant:
            self.placeIsSetEnPassant = True
            return self.placeEnPassant

        return None

    # get the position of the opponent enpassant is bieng performed on so opponent can be removed
    def getEnPassantOpponentPlace(self):
        if self.opponentPlaceIsSetEnPassant:
            self.opponentPlaceIsSetEnPassant = False
            return self.opponentPlaceEnPassant

        return None

    # gets all the places that a pawn can move
    def getPawnMovablePlaces(self, x, y):
        matrix = self.live_chessboard_matrix
        placeIds = ["" for a in range(4)]
        if x >= 0 and x <= 7 and y >= 0 and y <= 7:
            rightFwd1Select = Select()
            rightFwd1Element = rightFwd1Select.selectFromParentId(
                matrix, self.id_gen(y+1, x-1))
            leftFwd1Select = Select()
            leftFwd1Element = leftFwd1Select.selectFromParentId(
                matrix, self.id_gen(y+1, x+1))
            fwd1Select = Select()
            fwd1Element = fwd1Select.selectFromParentId(
                matrix, self.id_gen(y+1, x))
            fwd2Select = Select()
            # print("get pawn movable places %s"%self.id_gen(y+2, x))
            fwd2Element = fwd2Select.selectFromParentId(
                matrix, self.id_gen(y+2, x))
        
            if leftFwd1Element != None:
                if leftFwd1Element.piece_id != None and leftFwd1Element != "":
                    leftFwd1Id = leftFwd1Element.piece_id

                    if self.isType(leftFwd1Id, "player_"):
                        placeIds[0] = leftFwd1Element.parent_id

            if rightFwd1Element != None:
                if rightFwd1Element.piece_id != None and rightFwd1Element != "":
                    rightFwd1Id = rightFwd1Element.piece_id
                    if self.isType(rightFwd1Id, "player_"):
                        placeIds[1] = rightFwd1Element.parent_id

            if fwd1Element != None:
                if fwd1Element.piece_id != None and fwd1Element.piece_id == "":
                    placeIds[2] = fwd1Element.parent_id

            if fwd2Element != None:
                if fwd2Element.piece_id != None and fwd2Element.piece_id == "":
                    placeIds[3] = fwd2Element.parent_id
        # print("Pawn %s"%placeIds)
        return placeIds

    #checks if a piece is bieng attacked by the pawn
    def attackingPlaces(self, x, y):
        matrix = self.live_chessboard_matrix
        attackingPawnPlaces = self.shrinkPawnArray(
            self.getPawnMovablePlaces(x, y), "checking")
        new_array = []
        # print("Pawn %s"%attackingPawnPlaces)
        return attackingPawnPlaces

    # no gaps in array so can be used for selection
    def shrinkPawnArray(self, array, mechanic_needed):
        matrix = self.live_chessboard_matrix
        new_array = []

        if mechanic_needed == "playing":

            if array[0] != None:
                if array[0] != "":
                    new_array.append(array[0])

            if array[1] != None:
                if array[1] != "":
                    new_array.append(array[1])

            if array[2] != None:
                if array[2] != "":
                    new_array.append(array[2])

            if array[3] != None:
                if array[3] != "":
                    new_array.append(array[3])

        else:
            if array[0] != None:
                if array[0] != "":
                    new_array.append(array[0])

            if array[1] != None:
                if array[1] != "":
                    new_array.append(array[1])

        return new_array

    # adds the empassant postion if available to the pawns possible moves    
    def enPassantMovement(self, compPawnStartingPositions, currentEnPassantPlaceId, enPassantOpponentLeft, enPassantOpponentRight, isEnPassant, currentEnPassantOpponentPlaceId, pawnArray, x, y):
        matrix = self.live_chessboard_matrix
        newArray = []
        pawnHasLeft = False
        pawnHasRight = False
        leftOfPawn = self.id_gen(y, x-1)
        rightOfPawn = self.id_gen(y, x+1)
        leftOfPawnSelect = Select()
        leftOfPawnElement = leftOfPawnSelect.selectFromParentId(
            matrix, leftOfPawn)
        rightOfPawnSelect = Select()
        rightOfPawnElement = rightOfPawnSelect.selectFromParentId(
            matrix, rightOfPawn)

        if leftOfPawnElement != None:
            if leftOfPawnElement.piece_id != None and leftOfPawnElement.piece_id != "":
                enPassantSpace = self.pawnReadyEnPassant(
                    compPawnStartingPositions,
                    currentEnPassantPlaceId,
                    leftOfPawnElement.piece_id,
                    leftOfPawn
                )

                if enPassantSpace != "":
                    pawnArray.append(enPassantSpace)
                    enPassantOpponentLeft = leftOfPawnElement.piece_id
                    currentEnPassantOpponentPlaceId = leftOfPawn
                    isEnPassant = True
                    self.opponentPlaceIsSetEnPassant = True
                    self.opponentPlaceEnPassant = leftOfPawn

        if rightOfPawnElement != None:
            if rightOfPawnElement.piece_id != None and rightOfPawnElement.piece_id != "":
                enPassantSpace = self.pawnReadyEnPassant(
                    compPawnStartingPositions,
                    currentEnPassantPlaceId,
                    rightOfPawnElement.piece_id,
                    rightOfPawn
                )

                if enPassantSpace != "":
                    pawnArray.append(enPassantSpace)
                    enPassantOpponentRight = rightOfPawnElement.piece_id
                    currentEnPassantOpponentPlaceId = rightOfPawn
                    isEnPassant = True
                    self.opponentPlaceIsSetEnPassant = True
                    self.opponentPlaceEnPassant = rightOfPawn

            return pawnArray

    # checks if the pawn is ready for enpassant
    def pawnReadyEnPassant(self, compPawnStartingPositions, currentEnPassantPlaceId, pieceId, placeId):
        newPlaceId = ""

        if self.isType(pieceId, "player_pawn"):
            for i in range(len(compPawnStartingPositions)):
                if self.isType(pieceId, str(i+1)):
                    posBefore = self.findPlaceCoordinates(
                        compPawnStartingPositions[i])
                    y = posBefore[0] + 1
                    x = posBefore[1]
                    posNow = self.findPlaceCoordinates(placeId)
                    nY = posNow[0] - 1
                    nX = posNow[1]
                    placeIdWithPosBefore = self.id_gen(y, x)
                    placeIdWithPosNow = self.id_gen(nY, nX)

                    if placeIdWithPosBefore == placeIdWithPosNow:
                        newPlaceId = placeIdWithPosNow
                        currentEnPassantPlaceId = placeIdWithPosNow
                        self.placeIsSetEnPassant = True
                        self.placeEnPassant = placeIdWithPosNow

        return newPlaceId
